### PR TITLE
Improve aggressiveness of dropBuffer calls caused by cuts

### PR DESCRIPTION
**Pull request by lihaoyi**
In principle, we should be able to call `dropBuffer` immediately after the `lhs` of `lhs ~/ rhs` succeeds, as the cut occurs "in between" the `lhs` and `rhs`. 

Currently, we only call `dropBuffer` after the `rhs` succeeds, as when we're evaluating the "current" `lhs ~/ rhs`, after `lhs` succeeds we decide whether or not to `dropBuffer` based on the *previous* existence of cut, ignoring the fact that a cut had just occurred. Thus, if we had `a ~/ b ~ c`, we would call `dropBuffer` after `b` completes, while we really want it to occur after `a`.

This diff makes the consideration of `dropBuffer` also consider whether the "current" `CustomSequence`/iteration-in-`Sequence.Flat` just performed a cut. Thus, given `a ~/ b ~ c`, we now call `dropBuffer` immediately after `a`, in addition to after `b`.

Also does some miscellaneous changes:

- Convert `LoggedDropsParserInput` to use `SortedSet` and give it a nice
  `toString`, so it's easier to debug things when they misbehave

- Add a unit test to verify that top-level `~`s call `dropBuffer`, even
  without any cuts. This behaves as expected right now, but good to have
  a dedicated test for it

- Add unit tests to verify the new behavior

Tests seem to pass https://travis-ci.org/lihaoyi/fastparse/builds/153880403, at least as-passing-as-they-currently-are

Review by @vovapolu

### CODE REVIEW

File: fastparse/shared/src/test/scala/fastparse/iterator/IteratorTests.scala [hunk 1]
```diff
@@ -11,17 +11,55 @@ object IteratorTests extends TestSuite {
...
+      val Parsed.Success(res, i) = p.parseInput(input)
+      // Make sure that we drop immediately at position 2, since that is where
+      // the cut has taken place, rather than at position 4 as we did earlier.
+      assert(input.drops == Set(2, 4))
+    }
+    'whitespaceImmediateCutDrop{
+      import fastparse.noApi._
+      val White = fastparse.WhitespaceApi.Wrapper{
+        import fastparse.all._
+        NoTrace(" ".? ~ " ".rep) // note that the whitespace delimiter has cut
```
Comment (vovapolu) [inline fastparse/shared/src/test/scala/fastparse/iterator/IteratorTests.scala]: Hmm, whitespace delimiter hasn't cut in this case

**Comment by vovapolu** [comment]
Yes, that's true. I've missed it when I was thinking about placement of `dropBuffers`.

**Comment by vovapolu** [comment]
@lihaoyi Great :) Let's merge and look how your unit-test PR is going to work with new `dropBuffer` improvements.

**Comment by lihaoyi** [comment]
FTR, this reduced the Max Buffer Size of the iterator parsers in our various benchmarks by small amounts:

Benchmark  | Before | After
-------------|---------|-----
ScalaParse  | 1587 | 1555
PythonParse | 2025 | 2006
ByteParse | 40 | 36

CSSParse wasn't affected; I guess we need to put some cuts in that.

These numbers are all for batch-size 1, i.e. the "exact" buffer size required, without any real buffering.

Relatively small improvement overall, but non-zero

**Pull request merged**